### PR TITLE
New: space-before-function-paren rule (fixes #2028)

### LINKED ISSUE #2028: Rename space-before-function-parentheses -> space-before-function-parens
Someone pointed out that this is the only rule where we force users to spell "parentheses" whereas every other rule we simply use "parens". Probably best to be consistent going into v1.0.0.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/9557596-rename-space-before-function-parentheses-space-before-function-parens?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by btmills**

### CODE REVIEW

**Comment by btmills** [comment]
Once the renamed rule is added, I'll add a note to #1898 to remove `space-before-function-parentheses`.

File: docs/rules/space-before-function-parentheses.md [hunk 1]
```diff
@@ -1,5 +1,7 @@
 # Require or disallow spaces before function parentheses (space-before-function-parentheses)
 
+**Deprecation notice**: This rule is deprecated, as it has been renamed to [space-before-function-parens](space-before-function-parens.md) for consistency with other rules' names, which used "parens" instead of "parentheses". This rule, by the old name, will be removed in ESLint v1.0. The new rule is identical in everything except name.
+
```
Comment (lo1tuma) [inline docs/rules/space-before-function-parentheses.md]: I think you should also update the deprecation warning of `space-after-function-name` which links to `space-before-function-parentheses`.

Comment (btmills) [inline docs/rules/space-before-function-parentheses.md]: Amended to update the reference in `generator-star-spacing` as well. A search through the codebase turned up no other references.

**Comment by btmills** [comment]
While it's being renamed, should the new name use the plural "parens" or singular "paren"? It's only checking before the opening paren and not the closing one.

**Comment by btmills** [comment]
Before this is merged, it will need to incorporate the work done in #2083.

**Comment by nzakas** [comment]
I suppose "paren" is more accurate. Fine by me.

**Comment by btmills** [comment]
Rebased and updated.

**Comment by nzakas** [comment]
Lgtm

**Pull request merged**